File: sample/py/01_mandelbrot.py
# 01: マンデルブロ集合を PNG 画像として出力するサンプルです。
# 将来のトランスパイルを意識して、構文はなるべく素直に書いています。

from time import perf_counter
import png_helper


def escape_count(cx: float, cy: float, max_iter: int) -> int:
    """1点 (cx, cy) の発散までの反復回数を返す。"""
    x: float = 0.0
    y: float = 0.0
    i: int = 0
    while i < max_iter:
        x2: float = x * x
        y2: float = y * y
        if x2 + y2 > 4.0:
            return i
        y = 2.0 * x * y + cy
        x = x2 - y2 + cx
        i = i + 1
    return max_iter


def color_map(iter_count: int, max_iter: int) -> tuple[int, int, int]:
    """反復回数を RGB に変換する。"""
    if iter_count >= max_iter:
        return (0, 0, 0)

    # 簡単なグラデーション（青系 -> 黄系）
    t: float = (iter_count * 1.0) / (max_iter * 1.0)
    r: int = int(255.0 * (t * t))
    g: int = int(255.0 * t)
    b: int = int(255.0 * (1.0 - t))
    return (r, g, b)


def render_mandelbrot(
    width: int,
    height: int,
    max_iter: int,
    x_min: float,
    x_max: float,
    y_min: float,
    y_max: float,
) -> bytearray:
    """マンデルブロ画像の RGB バイト列を生成する。"""
    pixels: bytearray = bytearray()

    y: int = 0
    while y < height:
        py: float = y_min + (y_max - y_min) * ((y * 1.0) / ((height - 1) * 1.0))

        x: int = 0
        while x < width:
            px: float = x_min + (x_max - x_min) * ((x * 1.0) / ((width - 1) * 1.0))
            it: int = escape_count(px, py, max_iter)
            r: int
            g: int
            b: int
            if it >= max_iter:
                r = 0
                g = 0
                b = 0
            else:
                t: float = (it * 1.0) / (max_iter * 1.0)
                r = int(255.0 * (t * t))
                g = int(255.0 * t)
                b = int(255.0 * (1.0 - t))
            pixels.append(r)
            pixels.append(g)
            pixels.append(b)
            x = x + 1

        y = y + 1

    return pixels


def run_mandelbrot() -> None:
    width: int = 800
    height: int = 600
    max_iter: int = 400
    out_path: str = "sample/out/mandelbrot_01.png"

    start: float = perf_counter()

    pixels: bytearray = render_mandelbrot(
        width,
        height,
        max_iter,
        -2.2,
        1.0,
        -1.2,
        1.2,
    )
    png_helper.write_rgb_png(out_path, width, height, pixels)

    elapsed: float = perf_counter() - start
    print("output:", out_path)
    print("size:", width, "x", height)
    print("max_iter:", max_iter)
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_mandelbrot()

File: sample/cpp/01_mandelbrot.cpp
#include "cpp_module/gc.h"
#include "cpp_module/png.h"
#include "cpp_module/py_runtime.h"
#include "cpp_module/time.h"
#include <algorithm>
#include <any>
#include <cstdint>
#include <fstream>
#include <ios>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <type_traits>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
using namespace pycs::gc;

long long escape_count(double cx, double cy, long long max_iter)
{
    double x = 0.0;
    double y = 0.0;
    auto __pytra_range_start_1 = 0;
    auto __pytra_range_stop_2 = max_iter;
    auto __pytra_range_step_3 = 1;
    if (__pytra_range_step_3 == 0) throw std::runtime_error("range() arg 3 must not be zero");
    for (auto i = __pytra_range_start_1; (__pytra_range_step_3 > 0) ? (i < __pytra_range_stop_2) : (i > __pytra_range_stop_2); i += __pytra_range_step_3)
    {
        double x2 = (x * x);
        double y2 = (y * y);
        if (((x2 + y2) > 4.0))
        {
            return i;
        }
        y = (((2.0 * x) * y) + cy);
        x = ((x2 - y2) + cx);
    }
    return max_iter;
}

tuple<long long, long long, long long> color_map(long long iter_count, long long max_iter)
{
    if ((iter_count >= max_iter))
    {
        return std::make_tuple(0, 0, 0);
    }
    double t = py_div(iter_count, max_iter);
    long long r = static_cast<long long>((255.0 * (t * t)));
    long long g = static_cast<long long>((255.0 * t));
    long long b = static_cast<long long>((255.0 * (1.0 - t)));
    return std::make_tuple(r, g, b);
}

string render_mandelbrot(long long width, long long height, long long max_iter, double x_min, double x_max, double y_min, double y_max)
{
    string pixels = py_bytearray();
    auto __pytra_range_start_4 = 0;
    auto __pytra_range_stop_5 = height;
    auto __pytra_range_step_6 = 1;
    if (__pytra_range_step_6 == 0) throw std::runtime_error("range() arg 3 must not be zero");
    for (auto y = __pytra_range_start_4; (__pytra_range_step_6 > 0) ? (y < __pytra_range_stop_5) : (y > __pytra_range_stop_5); y += __pytra_range_step_6)
    {
        double py = (y_min + ((y_max - y_min) * py_div(y, (height - 1))));
        auto __pytra_range_start_7 = 0;
        auto __pytra_range_stop_8 = width;
        auto __pytra_range_step_9 = 1;
        if (__pytra_range_step_9 == 0) throw std::runtime_error("range() arg 3 must not be zero");
        for (auto x = __pytra_range_start_7; (__pytra_range_step_9 > 0) ? (x < __pytra_range_stop_8) : (x > __pytra_range_stop_8); x += __pytra_range_step_9)
        {
            double px = (x_min + ((x_max - x_min) * py_div(x, (width - 1))));
            long long it = escape_count(px, py, max_iter);
            long long r;
            long long g;
            long long b;
            if ((it >= max_iter))
            {
                r = 0;
                g = 0;
                b = 0;
            }
            else
            {
                double t = py_div(it, max_iter);
                r = static_cast<long long>((255.0 * (t * t)));
                g = static_cast<long long>((255.0 * t));
                b = static_cast<long long>((255.0 * (1.0 - t)));
            }
            pixels.push_back(r);
            pixels.push_back(g);
            pixels.push_back(b);
        }
    }
    return pixels;
}

void run_mandelbrot()
{
    long long width = 800;
    long long height = 600;
    long long max_iter = 400;
    string out_path = "sample/out/mandelbrot_01.png";
    double start = perf_counter();
    string pixels = render_mandelbrot(width, height, max_iter, (-2.2), 1.0, (-1.2), 1.2);
    pycs::cpp_module::png::write_rgb_png(out_path, width, height, pixels);
    double elapsed = (perf_counter() - start);
    py_print("output:", out_path);
    py_print("size:", width, "x", height);
    py_print("max_iter:", max_iter);
    py_print("elapsed_sec:", elapsed);
}

int main()
{
    run_mandelbrot();
    return 0;
}
